Limit number of chunks produced by iteration_shape

biddisco opened the pull request:
During testing, I found that parallel::for() was regularly producing around 5000 chunks in release mode and up to 100,000 chunks in debug mode (especially when stepping through with a debugger) as the chunker uses the time for an iteration to decide how big/small to make chunks (and god forbid that the kernel suspends your thread during that initial check).

This quick fix, limits the chunk size by limiting the number of chunks to 4xCores. It's a small number and there's a strong argument that it should be bigger then 4, but unless the user is running a lot of long running tasks and startving the system of others, it should improve performance overall. I saw a noticable improvement after adding this.

I may experiment with some new parameters for the executor policy that would allow chunk_resolution::micro/millseconds or chunk_resolution::block_size using some unit checks. This would alow the user to place a resolution of (say) 100microseconds per task, or a chunk size, but then we'd also need a maximum_chunks parameter default that would limit the extreme cases when a test takes seconds and the 100useconds limit triggers 100,000 chunks ...

hkaiser commented:
LGTM, thanks!

sithhell commented:
Would be nice if we could tune this parameter through a define somehow, that would allow us to experiment with it a bit more.

biddisco commented:
Making it a #define would be easy, but my suspicion is that there are too many #define type options that the average user would get scared by. A -Ihpx.default.chunk.factor=10 type approach would be more appealing to me. I would like to tweak the executor parameters to allow it to be specified the same way the static_chunk_size is done (as mentioned earlier).

Don't merge this yet. I may have a go at the parameters ...

hkaiser commented:
> A -Ihpx.default.chunk.factor=10 type approach would be more appealing to me. 

Except for the parameter name, I'd agree. 'chunk.factor' of what? Let's try to be more descriptive and unambiguous.

hkaiser commented:
@biddisco I removed the hard-coded limit and replaced it by a call into the executor parameters object allowing for the user to easily customize the settings. I left the default value at `4 * cores` though.

biddisco commented:
Having given this a look. I'm not sure that `maximal_size_of_chunks` is the right name. If I loop over 1000000 iterations and the maximal chunk size is 500000, then I get two chunks, which is fine. But the code is actually setting maximal size of chunks to `cores*4` which I think is really the `maximal_number_of_chunks` -. should I change the name?

biddisco commented:
Also, there are times when an iteration is genuinely expensive and might take seconds or even minutes - ideally these tasks should be broken down into smaller ones, but that might require more work - so we should also provide a `maximal_time_per_chunk` parameter that would default to something like 100ms. This would allow the user to sets a chunk size either based on number of iterations (or cores in our case), or based on time when you do not want to have all threads running tasks that take seconds and block all other tasks from running.

hkaiser commented:
@biddisco 

> ...But the code is actually setting maximal size of chunks to  cores*4  which I think is really the  `maximal_number_of_chunks` -. should I change the name?

Uhh, silly me. What was I thinking? Thanks for pointing this out.

> ...  so we should also provide a  `maximal_time_per_chunk`  parameter ...

In some sense this is already available (see: https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/parallel/executors/auto_chunk_size.hpp#L54). I'm open to discussions, though.

biddisco commented:
aha. I didn't realize the auto chunk size had a time param - then that's fine.

Are you fixing the maximal_number_of_chunks or am I?

hkaiser commented:
> Are you fixing the maximal_number_of_chunks or am I?

Let me do it. You have other things on your plate.

biddisco commented:
@hkaiser I rebased this onto master and added one small fix. I think the chunk size calculation was incorrect after the changes. If all looks ok, please feel free to merge whenver you wish/can.

hkaiser commented:
Thanks John!
